### ISSUE TITLE
augur parse silently drops metadata from records with duplicate strain/name fields

### ISSUE THREAD

**Issue by huddlej** [nextstrain/augur#616]
**Current Behavior**

augur parse reads in sequence data, parses metadata from FASTA headers, and [indexes each sequence by the name field it finds in the FASTA headers](https://github.com/nextstrain/augur/blob/efcc8be3bcf72ab4beb237afcc5b4b002bbaa806/augur/parse.py#L123). When the input data have different records with the same strain name, this behavior causes:

* all duplicate sequences to be written out in the FASTA output
* all but the last of the duplicate records to be dropped from the metadata (without any notice to the user)

In contrast, augur align [defines its own read_sequences function that attempts to identify duplicates](https://github.com/nextstrain/augur/blob/efcc8be3bcf72ab4beb237afcc5b4b002bbaa806/augur/align.py#L178-L192) (same name, different sequences).

**Expected behavior**

augur parse should throw an error when duplicate strain names are detected and alert the user about which specific strains are duplicates and how many there are.

Additionally, it would be nice if augur parse allowed the user to define which field(s) to use as a unique identifier. In the case above, this option would allow users to key sequences by the unique `accession` field instead of the non-unique `strain` field.

**How to reproduce**

Install augur v10.0.2 and [download attached zip file of sequences](https://github.com/nextstrain/augur/files/5228387/minimal_augur_parse_data.fasta.zip). This subset comes from [real data](https://elifesciences.org/download/aHR0cHM6Ly9jZG4uZWxpZmVzY2llbmNlcy5vcmcvYXJ0aWNsZXMvMzEyNTcvZWxpZmUtMzEyNTctZmlnMS1kYXRhNS12My56aXA-/elife-31257-fig1-data5-v3.zip?_hash=YhuQfm%2BGO%2BY6MsWLZB4WrPQvYtSlHOhLnzwnvTaesws%3D) from [Dudas et al. 2018](https://elifesciences.org/articles/31257). Then run the following commands to verify the issue:

```
# minimal parse
$ augur parse \
    --sequences minimal_augur_parse_data.fasta \
    --fields strain accession host date \
    --output-sequences sequences.fasta \
    --output-metadata metadata.tsv

# confirm sequences are all present
$ grep "^>" sequences.fasta
>Jeddah-1
>Jeddah-1

# confirm that metadata only exists for one record
$ cat metadata.tsv
strain	accession	host	date
Jeddah-1	KF958702	human	2013-11-05
```

**Possible solution**

1. Check for presence of duplicate strain names in a first pass through the input records. Collect all metadata for these duplicates. If this list of duplicates isn't empty after the first pass, emit the metadata details to the user on stderr and exit with an error.

2. Allow users to provide a `--record-id` or `--sequence-id` argument that specifies which column of the metadata to use as a key.

**Your environment: if running Nextstrain locally**  
 - Operating system: OS X
 - Version: augur 10.0.2

**Comment by avrajit** [comment]
Hi I am also getting the same problem. What is the solution? Do I just copy your scripts and replace it with old augur one?

**Comment by huddlej** [comment]
Hi @avrajit, we haven't worked out a general solution to this issue yet, but we do have some scripts in our SARS-CoV-2 workflow that attempt to resolve duplicate metadata and sequence records. For example, [our sanitize metadata script](https://github.com/nextstrain/ncov/blob/71508e0a31935c0682f2ad4b17ce862f2328c229/workflow/snakemake_rules/main_workflow.smk#L1-L24) resolves duplicates by selecting the record with the latest GISAID accession and [our sanitize sequences script](https://github.com/nextstrain/ncov/blob/71508e0a31935c0682f2ad4b17ce862f2328c229/workflow/snakemake_rules/main_workflow.smk#L81-L84) selects the first instance of a duplicate sequence for a given strain.

If you're not working with SARS-CoV-2, you may still be able to copy and modify these scripts to work for your analysis.

**Comment by huddlej** [comment]
Revisiting this issue a couple years later, my thinking about the best solution has changed. We should try to keep Augur commands as tightly scoped as possible to solve specific problems. Resolving duplicates should be outside the responsibilities of the `parse` command; another Augur command should be responsible for deduplicating/sanitizing sequences and metadata.

Based on this thinking, a better solution would be:

1. Replace the metadata dictionary structure in `parse` with a list of metadata dictionaries (e.g., `metadata_records`).
2. Create the metadata data frame from the list of dictionaries with `pd.DataFrame(metadata_records)` instead of using `from_dict`.

We could optionally check for duplicate strain ids in the metadata and print a warning, but I don't think `parse` should throw an error when duplicates exist.

**Comment by emmahodcroft** [comment]
Funnily, I ran into a duplicate strain issue yesterday only after getting to tree-building (can't remember if it was `tree` or `refine`), and it was annoying - I don't use `parse` but have a number of pre-tree-building steps, and I wish I'd detected the duplicate earlier to save having to backtrack.
While I agree that things should be tightly scoped, for most tree-building you need to have unique identifiers, and it's definitely easier to find out about things that violate this earlier rather than later. I wonder if a middle-ground might be changing the data structures as you suggest, but by default erroring for duplicates - which you can override with a flag if you want. Since `parse` is likely very early in the workflow, this would alert users to the issue quickly - and they can either go with it (perhaps if they are not proceeding to tree-building) by changing a flag, or resolve the issue before they hit errors in `tree`/`refine` which may be many steps down the line?

**Comment by huddlej** [comment]
That's a good idea for an intermediate solution to error on duplicates by default. I was thinking that we need a new command like `augur deduplicate` or `augur sanitize` that would handle duplicates like the "sanitize" scripts in ncov do now.

I would expect that few Nextstrain users touch `augur parse`, since it's not needed for SC2 work, but it is definitely needed for working with GISAID flu data. Maybe more people will start using it, but I don't think we can expect everyone to start their workflow from that command.

**Comment by emmahodcroft** [comment]
Yes, true - perhaps `augur parse` isn't a good place to put such a check just because few people use it - it really only works if you have nicely structured sequence names (and if you're downloading from Genbank you mostly pray they're unique and don't have a weird special character in them).
Something like `augur sanitize` might be a nice idea with the idea that it's lightweight, generic, and that most people incorporate it into their workflows. It could check for duplicates, maybe check for special symbols, new lines, etc, try to format dates?

A while back a PhD student of Richard worked on a neat script that also allowed you to interactively standardize metadata, including get lat-longs (like we do for `ncov-ingest` curation - we reused that old logic for this!) and standardize locations. That's probably best as a separate function, since it's more heavy and not everyone will need it or want it -- but I'd forgotten the idea until now. Perhaps as the pandemic wanes, could come back to it.
